### PR TITLE
implement new derive resource id from webb proposal

### PR DESCRIPTION
**Pull request by dharjeezy**
Update resource ID in XAnchor with target chain ID and target tree ID according to #183

### CODE REVIEW

**Review by drewstone — CHANGES_REQUESTED** [review]
Looks OK but this doesn't not solve the issue we started with @dharjeezy. We need to ensure that the chain IDs are the _typed chain ID_ (u64 variant) when we store them in EdgeMetadata and inside the chain's storage.

We also need to ensure that proposal's resource IDs contain the destination chain. This means the chain in the resource ID should be the chain where the proposal is being _executed_.

File: pallets/xanchor/src/lib.rs [hunk 1]
```diff
@@ -464,10 +467,11 @@ pub mod pallet {
 			);
 			// Now we can remove the pending linked anchor.
 			PendingLinkedAnchors::<T, I>::remove(caller_chain_id, my_tree_id);
-			let r_id = utils::derive_resource_id(
+			let r_id = utils::derive_resource_id_v2(
 				caller_chain_id.try_into().unwrap_or_default(),
```
[the remark this answers is not in the document]
Comment (dharjeezy) [reply pallets/xanchor/src/lib.rs:471]: Alright

File: pallets/xanchor/src/lib.rs [hunk 2]
```diff
@@ -508,24 +512,28 @@ pub mod pallet {
...
+
+			let anchor_update_proposal = AnchorUpdateProposal::from(anchor_update_proposal_bytes);
+
+			let (tree_id, r_chain_id) = utils::parse_resource_id_v2::<T::TreeId, T::ChainId>(
+				anchor_update_proposal.header().resource_id(),
 			);
+
+			//construct the metadata
+			let metadata = EdgeMetadata {
+				src_chain_id: r_chain_id,
```
[the remark this answers is not in the document]
Comment (dharjeezy) [reply pallets/xanchor/src/lib.rs:531]: it is a u32. I will go ahead and make it a u64 with your comments. `u64::from_be_bytes([typed_chain_id.to](http://typed_chain_id.to/)_bytes)`

File: pallets/xanchor/src/lib.rs [hunk 3]
```diff
@@ -639,13 +648,30 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
...
-			src_chain_id: para_id_to_chain_id::<T, I>(my_para_id),
-			//src_chain_id: my
-			root,
-			latest_leaf_index,
-		};
+		let src_chain_id = u32::from(my_para_id);
+
+		// construct resource id
+		let resource_id =
+			utils::derive_resource_id_v2(src_chain_id, tree_id.try_into().unwrap_or_default());
```
[the remark this answers is not in the document]
Comment (dharjeezy) [reply pallets/xanchor/src/lib.rs:655]: Alright

File: pallets/xanchor/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dharjeezy) [reply pallets/xanchor/src/lib.rs:471]: Hello @drewstone for here, i think this is correct, as the target chain id is also the same as the caller chain id.

File: pallets/xanchor/src/lib.rs [hunk 4]
```diff
@@ -508,24 +512,43 @@ pub mod pallet {
...
+
+			let typed_chain_id_of_caller: u64 =
+				utils::get_typed_chain_id_in_u64(r_chain_id.try_into().unwrap_or_default());
+
+			// verify that the chain id and the one from the parsed resource id matches
+			ensure!(my_chain_id == typed_chain_id_of_caller, Error::<T, I>::InvalidPermissions);
+
+			//construct the metadata
+			let metadata = EdgeMetadata {
+				src_chain_id: caller_chain_id,
```
Comment (drewstone) [inline pallets/xanchor/src/lib.rs:546]: The chain id in the edge metadata should be the typed chain id (u64)

**Review by drewstone — CHANGES_REQUESTED** [review]

File: pallets/xanchor/src/lib.rs (hunk 4 above)
Comment (dharjeezy) [reply pallets/xanchor/src/lib.rs:546]: Yes. It is the typed chain id in u64 you can take a look at line 541 to see how i derived it(`caller_chain_id`).

**Comment by drewstone** [comment]
Looks good so far, need to cleanup the print statements and merge master to update the link proposal functionality.

**Pull request merged**